src/system/process.py

```diff
@@ -0,0 +1,71 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+import logging
+import subprocess
+import tempfile
+
+import psutil  # type: ignore
+
+
+class Process:
+    def __init__(
+        self
+    ):
+        self.process = None
+        self.stdout = None
+        self.stderr = None
+
+    def start(self, command, cwd):
+        self.stdout = tempfile.NamedTemporaryFile()
+        self.stderr = tempfile.NamedTemporaryFile()
+
+        self.process = subprocess.Popen(
+            command,
+            cwd=cwd,
+            shell=True,
+            stdout=self.stdout,
+            stderr=self.stderr,
+        )
+
+    def terminate(self):
+        parent = psutil.Process(self.process.pid)
+        logging.debug("Checking for child processes")
+        child_processes = parent.children(recursive=True)
+        for child in child_processes:
+            logging.debug(f"Found child process with pid {child.pid}")
+            if child.pid != self.process.pid:
+                logging.debug(f"Sending SIGKILL to {child.pid} ")
+                child.kill()
+        logging.info(f"Sending SIGKILL to PID {self.process.pid}")
+        
+        self.process.kill()
+    
+        logging.info(f"Process terminated with exit code {self.process.returncode}")
+        if self.stdout:
```

Comment:
Does `self.stdout_data = self.stdout.read()` not work?

Reply:
Just want to leverage the benefit of `with` statement to not worry about explicitly calling close. 

https://www.geeksforgeeks.org/with-statement-in-python/

Reply:
I need to revert the change for using `with`. The windows check is failing.

```
        if self.stdout:
>           with open(self.stdout.name) as stdout:
E           PermissionError: [Errno 13] Permission denied: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpxl83o9tt'

src\system\process.py:59: PermissionError

----------- coverage: platform win32, python 3.7.9-final-0 -----------
Coverage XML written to file coverage.xml

=========================== short test summary info ===========================
FAILED tests/tests_system/test_process.py::TestProcess::test - PermissionErro...
1 failed, 355 passed in 61.81s (0:01:01)
```

I will just call read() and then close()

Reply:
i don't think this is the problem.

Reply:
After digging into the tempfile doc https://docs.python.org/3/library/tempfile.html, opening for a second time has different behaviors in different platforms.

```
Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later). 
```